### PR TITLE
Pc/present cv

### PR DESCRIPTION
**Pull request by Blaizzy**
For the How to track and present cross validation results in Neptune guide

### CODE REVIEW

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 1]
```diff
@@ -0,0 +1,583 @@
+{
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:1]: Pls make sure that you submit clean notebooks  -> no outputs

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+            "    Uninstalling future-0.16.0:\n",
+            "      Successfully uninstalled future-0.16.0\n",
+            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
+            "datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.\u001b[0m\n",
+            "Successfully installed GitPython-3.1.24 PyJWT-2.3.0 boto3-1.19.4 botocore-1.22.4 bravado-11.0.3 bravado-core-5.17.0 future-0.18.2 gitdb-4.0.9 jmespath-0.10.0 jsonref-0.2 monotonic-1.6 neptune-client-0.13.0 rfc3987-1.3.8 s3transfer-0.5.0 simplejson-3.17.5 smmap-5.0.0 strict-rfc3339-0.7 swagger-spec-validator-2.7.3 urllib3-1.25.11 webcolors-1.11.1 websocket-client-1.2.1\n"
+          ]
+        }
+      ],
+      "source": [
+        "! pip install neptune-client numpy==1.19.5 torch torchvision sklearn"
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:142]: only numpy has version, maybe you can drop it?

Comment (kamil-kaczmarek) [reply how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:142]: or use numpy==1.21.0 -> security reasons

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "n1NVL5h6MlLq",
+        "tags": []
+      },
+      "outputs": [],
+      "source": [
+        "import neptune.new as neptune\n",
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:163]: make sure that you sort imports properly -> PEP 8 -- Style Guide for Python Code -> https://www.python.org/dev/peps/pep-0008/#imports

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
+# How to present CV with Neptune
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:1]: you can drop this comment from line 1

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+import neptune.new as neptune
+import torch
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:3]: Pls use use black formatting in both colab and script.

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import torch.nn as nn
+import torch.optim as optim
+from torchvision import datasets, transforms
+from torch.utils.data import SubsetRandomSampler, DataLoader
+from sklearn.model_selection import KFold
+from statistics import mean
+
+# Step 1: Create a Neptune Run
+run = neptune.init(
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:12]: Please, keep order of arguments as they appear in the docs -> standardization.
https://docs.neptune.ai/api-reference/neptune#.init_run

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    project="common/showroom", tags="cross-validation", api_token="ANONYMOUS"
+)
+
+# Step 2: Log config and hyperparameters
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:16]: Here, you did not log it, just defined dict.

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+parameters = {
+    "epochs": 1,
+    "lr": 1e-2,
+    "bs": 10,
+    "input_sz": 32 * 32 * 3,
+    "n_classes": 10,
+    "k_folds": 2,
+    "model_name": "checkpoint.pth",
+    "device": torch.device("cuda:0" if torch.cuda.is_available() else "cpu"),
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:25]: pls not mix defining parameter with computing it. device methods should be outside dict.

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            transforms.RandomHorizontalFlip(),
+            transforms.ToTensor(),
+            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
+        ]
+    )
+}
+trainset = datasets.CIFAR10(data_dir, transform=data_tfms["train"], download=True)
+dataset_size = len(trainset)
+
+run["global/dataset/CIFAR-10"].track_files(data_dir)
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:73]: nice!

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        epoch_loss += loss.item() * x.size(0)
+    epoch_acc_list.append((epoch_acc / len(train_loader.sampler)) * 100)
+    epoch_loss_list.append(epoch_loss / len(train_loader.sampler))
+     
+    # Log model checkpoint       
+    torch.save(model.state_dict(), f"./{parameters['model_name']}")
+    run[f'fold_{fold}/checkpoint'].upload(parameters['model_name'])
+    
+run["global/metrics/train/mean_acc"] = mean(epoch_acc_list)
+run["global/metrics/train/mean_loss"] = mean(epoch_loss_list)
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:110]: new line at the end of file -> black formatting will handle that

File: how-to-guides/present-cross-validation-results/scripts/run_examples.sh [hunk 11]
```diff
@@ -0,0 +1,5 @@
+echo "Installing requirements..."
+pip install -r requirements.txt
+
+echo "Running neptune_cross_valition.py ..."
+python neptune_cross_validation.py
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/run_examples.sh:5]: new line at the end

**Review by kamil-kaczmarek — CHANGES_REQUESTED** [review]

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb (hunk 1 above)
Comment (Blaizzy) [reply how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:1]: Done ✅

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb (hunk 2 above)
Comment (Blaizzy) [reply how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:29]: I will drop numpy has it comes with torchvision and I will version all but neptune-client.

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py (hunk 5 above)
Comment (Blaizzy) [reply how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:2]: Done ✅

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py (hunk 4 above)
Comment (Blaizzy) [reply how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py]: Done ✅

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py (hunk 7 above)
Comment (Blaizzy) [reply how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:15]: I did. It's right after it.

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py (hunk 8 above)
Comment (Blaizzy) [reply how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:24]: I think of the device as a parameter to be logged so you easily re-instantiate your model into the right device in case you want to reproduce an experiment as you would do with the seed value. 

But this is only an example so it's fine to me to leave it out 👍

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "JhE6C9xv-lXK"
+   },
+   "source": [
+    "# Stop run\n",
+    "\n",
+    "<font color=red>**Warning:**</font><br>\n",
+    "Once you are done logging, you should stop tracking the run using the `stop()` method.\n",
+    "This is needed only while logging from a notebook environment. While logging through a script, Neptune automatically stops tracking once the script has completed execution."
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:394]: Maybe you can add one more very short section that tell user to go to the link with previously created run and explore results? (You can check our colab template notebook for reference).

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "run.stop()"
+   ]
+  }
+ ],
+ "metadata": {
+  "accelerator": "TPU",
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:415]: TPU accelerator, fancy :wink:

**Review by kamil-kaczmarek — CHANGES_REQUESTED** [review]

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+    "**(Optional)** To log to your Neptune project:\n",
+    "\n",
+    "* [Create a Neptune account](https://app.neptune.ai/register/)\n",
+    "\n",
+    "* [Find your API token](https://docs.neptune.ai/getting-started/installation#authentication-neptune-api-token)\n",
+    "* [Find your project name](https://docs.neptune.ai/getting-started/installation#setting-the-project-name)\n",
+    "\n",
+    "Pass your credentials to project and api_token arguments of neptune.init()\n",
+    "\n",
+    "`run = neptune.init(api_token='<YOUR_API_TOKEN>', project='<YOUR_WORKSPACE/YOUR_PROJECT>')` # pass your credentials\n"
```
[the remark this answers is not in the document]
Comment (kamil-kaczmarek) [reply how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:81]: You can check our colab template notebook for reference.

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+# Step 2: Log config and hyperparameters
+parameters = {
+    "epochs": 1,
+    "lr": 1e-2,
+    "bs": 10,
+    "input_sz": 32 * 32 * 3,
+    "n_classes": 10,
+    "k_folds": 2,
+    "model_name": "checkpoint.pth",
+    "seed": 42,
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:28]: Use full names, that is:
lr => learning_rate
bs -> batch_size
etc.

File: how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+            transforms.RandomHorizontalFlip(),
+            transforms.ToTensor(),
+            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
+        ]
+    )
+}
+trainset = datasets.CIFAR10(data_dir, transform=data_tfms["train"], download=True)
+dataset_size = len(trainset)
+
+run["global/dataset/CIFAR-10"].track_files(data_dir)
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:77]: We should use sth. else instead of "global",

How would you call it? Maybe "Training_results", or "Results"?

Comment (kamil-kaczmarek) [reply how-to-guides/present-cross-validation-results/scripts/neptune_cross_validation.py:77]: Make sure to chage it also in notebook

**Review by kamil-kaczmarek — CHANGES_REQUESTED** [review]
@Blaizzy 
Few changes and its ready

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 17]
```diff
@@ -0,0 +1,496 @@
...
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "54CFqKRANLK5",
+    "outputId": "a19f2835-22ab-41c9-ab13-e6e4d8fdbea9",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "! pip install neptune-client torch==1.10.2 torchvision==0.11.3 scikit-learn==1.0.2 #add other dependencies"
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:61]:
```suggestion
    "! pip install neptune-client torch==1.10.2 torchvision==0.11.3 scikit-learn==1.0.2"
```

Comment (kamil-kaczmarek) [reply how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:61]: All dependencies should be provided here

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb [hunk 18]
```diff
@@ -0,0 +1,496 @@
...
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "When training models with cross-validation, you can use Neptune namespaces to organize, visualize and compare models.\n",
+    "\n",
+    "By the end of this guide, you will learn how to organize your run to track cross-validation metadata, so that you can easily analyze the results.\n",
+    "\n",
+    "[See this example in Neptune](https://app.neptune.ai/o/common/org/showroom/e/SHOW-3700)\n",
+    "\n",
+    "[![image](https://files.gitbook.com/v0/b/gitbook-x-prod.appspot.com/o/spaces%2F-MT0sYKbymfLAAtTq4-t%2Fuploads%2FLRq4l78xEp31FFS5WzFM%2Fimage.png?alt=media&token=be868c83-7b1b-4314-b138-677531720f0e)](https://app.neptune.ai/o/common/org/showroom/e/SHOW-3700)\n",
```
Comment (kamil-kaczmarek) [inline how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:24]: This screenshot is almost empty -> too much of a whitespace. Pls adapt it + pls make screenshots with zoomed in view.

**Review by kamil-kaczmarek — CHANGES_REQUESTED** [review]

**Comment by kamil-kaczmarek** [comment]
@Blaizzy 

**Generic comment about Notebook.**

We just merged style guide 1.2. Have a look at [Monitor_ML_runs_live.ipynb](https://github.com/neptune-ai/examples/blob/main/how-to-guides/monitor-ml-runs/notebooks/Monitor_ML_runs_live.ipynb). You can almost copy&paste these sections (subject to copy fixes to fit your CV how to):
1. Before you start
1. Install Neptune and dependencies
1. Initialize Neptune and create new run

**5 fold CV**
[Example run](https://app.neptune.ai/o/common/org/showroom/e/SHOW-3700/all) has two folds, while screenshot has 5. Five folds are more realistic, pls adapt the code and example run to have 5 folds.

File: how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb (hunk 17 above)
Comment (Blaizzy) [reply how-to-guides/present-cross-validation-results/notebooks/Neptune_cross_validation.ipynb:61]: @kamil-kaczmarek As far as I know, this project has no missing dependencies.

If there are any specific dependencies I'm missing that you identified, please send them through, and I will add them right away.

**Comment by Blaizzy** [comment]
@kamil-kaczmarek all changes are done

**Review by kamil-kaczmarek — APPROVED** [review]
LGTM !
